File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/a6cf6bbb24a3292117b2614e99ecc5b8.java
import java.util.*;
import java.io.*;

public class a6cf6bbb24a3292117b2614e99ecc5b8_nc {
    static final long MOD = 1000000007;
    static long[] pos, w;
    static TreeSegment<Long, Long, Long, Long, Long> ts, resp;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int n = Integer.parseInt(st.nextToken());
        int q = Integer.parseInt(st.nextToken());
        
        pos = new long[n];
        w = new long[n];
        
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            pos[i] = Long.parseLong(st.nextToken());
        }
        
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            w[i] = Long.parseLong(st.nextToken());
        }

        ts = new TreeSegment<>(w, (a, b) -> a + b, (value, mod, tl, tr) -> mod, (oldVal, newVal, tl, tr) -> newVal, 1337L);
        
        long[] subw = new long[n];
        for (int i = 0; i < n; i++) {
            subw[i] = (w[i] * (i - pos[i])) % MOD;
        }

        resp = new TreeSegment<>(subw, (a, b) -> (a + b) % MOD, (value, mod, tl, tr) -> mod, (oldVal, newVal, tl, tr) -> newVal, 1337L);

        for (int i = 0; i < q; i++) {
            st = new StringTokenizer(br.readLine());
            long x = Long.parseLong(st.nextToken());
            long y = Long.parseLong(st.nextToken());

            if (x < 0) {
                int ind = (int) -x - 1;
                long newVal = y;
                ts.update(ind, ind + 1, newVal);
                resp.update(ind, ind + 1, (newVal * (ind - pos[ind])) % MOD);
            } else {
                int l = (int) x - 1;
                int r = (int) y;

                long sm = ts.query(l, r);

                int bestL = ts.findRight(l, sm, 0);

                long xVal = pos[bestL] - bestL + l;

                long left = 0;
                if (l < bestL) {
                    left = (resp.query(l, bestL) + ts.query(l, bestL) * (xVal - l)) % MOD;
                }
                long right = 0;
                if (bestL < r) {
                    right = (resp.query(bestL, r) + ts.query(bestL, r) * (xVal - l)) % MOD;
                }

                System.out.println(floorMod(left - right, MOD));
            }
        }
    }

    static long floorMod(long a, long b) {
        if (a % b == 0) return 0;
        if (a >= 0 && b >= 0) return a % b;
        if (a <= 0 && b <= 0) return a % b;
        return Math.abs(b) - (Math.abs(a) % Math.abs(b));
    }

    static class TreeSegment<Value, Mod, Magic, ApplyMod, MergeMods> {
        static class Node {
            boolean haveMod = false;
            long value;
            long mod;
        }

        Magic magic;
        ApplyMod applyMod;
        MergeMods mergeMods;
        Node[] S;
        int n;

        public TreeSegment(long[] src, Magic magic, ApplyMod applyMod, MergeMods mergeMods, long def) {
            this.magic = magic;
            this.applyMod = applyMod;
            this.mergeMods = mergeMods;
            init(src);
        }

        void init(long[] src) {
            n = src.length;
            S = new Node[2 * n + 1];
            for (int i = 0; i < S.length; i++) {
                S[i] = new Node();
            }
            init(0, 0, n, src);
        }

        void init(int i, int tl, int tr, long[] src) {
            if (tl + 1 == tr) {
                S[i].value = src[tl];
            } else {
                int mid = (tl + tr) / 2;
                init(i + 1, tl, mid, src);
                init(i + (mid - tl) * 2, mid, tr, src);
                recalcValue(i, tl, tr);
            }
        }

        void recalcValue(int i, int tl, int tr) {
            if (tl + 1 != tr) {
                int mid = (tl + tr) / 2;
                S[i].value = magic.apply(S[i + 1].value, S[i + (mid - tl) * 2].value);
            }
        }

        void push(int i, int tl, int tr) {
            if (S[i].haveMod) {
                S[i].value = applyMod.apply(S[i].value, S[i].mod, tl, tr);
                if (tl + 1 != tr) {
                    int mid = (tl + tr) / 2;
                    addMod(i + 1, tl, mid, S[i].mod);
                    addMod(i + (mid - tl) * 2, mid, tr, S[i].mod);
                }
                S[i].haveMod = false;
            }
        }

        void addMod(int i, int tl, int tr, long sub) {
            if (S[i].haveMod) {
                S[i].mod = mergeMods.apply(S[i].mod, sub, tl, tr);
            } else {
                S[i].mod = sub;
                S[i].haveMod = true;
            }
        }

        long query(int ql, int qr) {
            assert(ql < qr);
            return __query(0, 0, n, ql, qr);
        }

        long __query(int i, int tl, int tr, int ql, int qr) {
            push(i, tl, tr);
            if (ql <= tl && tr <= qr) {
                return S[i].value;
            } else {
                int mid = (tl + tr) / 2;
                if (intersects(tl, mid, ql, qr) && intersects(mid, tr, ql, qr)) {
                    return magic.apply(__query(i + 1, tl, mid, ql, qr), __query(i + (mid - tl) * 2, mid, tr, ql, qr));
                } else if (intersects(tl, mid, ql, qr)) {
                    return __query(i + 1, tl, mid, ql, qr);
                } else {
                    return __query(i + (mid - tl) * 2, mid, tr, ql, qr);
                }
            }
        }

        boolean intersects(int tl, int tr, int ql, int qr) {
            return !(tr <= ql || qr <= tl);
        }

        void update(int ql, int qr, long mod) {
            __update(0, 0, n, ql, qr, mod);
        }

        void __update(int i, int tl, int tr, int ql, int qr, long mod) {
            push(i, tl, tr);
            if (ql <= tl && tr <= qr) {
                addMod(i, tl, tr, mod);
            } else {
                int mid = (tl + tr) / 2;
                if (intersects(tl, mid, ql, qr)) {
                    __update(i + 1, tl, mid, ql, qr, mod);
                }
                if (intersects(mid, tr, ql, qr)) {
                    __update(i + (mid - tl) * 2, mid, tr, ql, qr, mod);
                }
                recalcValue(i, tl, tr);
            }
        }

        int findRight(int L, long sm, long start) {
            if (sm >= start) return L - 1;
            int rez = __findRight(0, 0, n, L, sm, start).first;
            return (rez == -1) ? n : rez;
        }

        Pair<Integer, Long> __findRight(int i, int tl, int tr, int L, long sm, long pref) {
            push(i, tl, tr);
            if (tr <= L) return new Pair<>(-1, pref);
            if (tl + 1 == tr) {
                long thisVal = magic.apply(pref, S[i].value);
                if (thisVal * 2 >= sm) {
                    return new Pair<>(tl, thisVal);
                } else {
                    return new Pair<>(-1, thisVal);
                }
            } else {
                int mid = (tl + tr) / 2;
                if (L <= tl) {
                    long left = magic.apply(pref, S[i + 1].value);
                    if (left * 2 >= sm) {
                        return __findRight(i + 1, tl, mid, L, sm, pref);
                    } else {
                        return __findRight(i + (mid - tl) * 2, mid, tr, L, sm, left);
                    }
                } else {
                    Pair<Integer, Long> t = __findRight(i + 1, tl, mid, L, sm, pref);
                    if (t.first != -1) return t;
                    return __findRight(i + (mid - tl) * 2, mid, tr, L, sm, t.second);
                }
            }
        }
    }

    static class Pair<F, S> {
        F first;
        S second;

        Pair(F first, S second) {
            this.first = first;
            this.second = second;
        }
    }
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/C/a6cf6bbb24a3292117b2614e99ecc5b8.c
#include <stdio.h>
#include <stdlib.h>

#define MAXMEM 200 * 1000 * 1024

char _memory[MAXMEM];
size_t _ptr = 0;

void* operator new(size_t _x) {
    _ptr += _x;
    if (_ptr >= MAXMEM) {
        exit(1);
    }
    return _memory + _ptr - _x;
}

void operator delete(void*) noexcept {}

typedef struct {
    int have_mod;
    int value;
    int mod;
} Node;

typedef struct {
    int n;
    Node* S;
} TreeSegment;

int floor_mod(int a, int b) {
    if (a % b == 0) {
        return 0;
    }
    if (a >= 0 && b >= 0) {
        return a % b;
    }
    if (a <= 0 && b <= 0) {
        return a % b;
    }
    return abs(b) - (abs(a) % abs(b));
}

int magic(int a, int b) {
    return a + b;
}

int apply_mod(int value, int mod, int tl, int tr) {
    return mod;
}

int merge_mods(int mod1, int mod2, int tl, int tr) {
    return mod2;
}

void init(TreeSegment* ts, int i, int tl, int tr, int* src) {
    if (tl + 1 == tr) {
        ts->S[i].value = src[tl];
    } else {
        int mid = (tl + tr) / 2;
        init(ts, i + 1, tl, mid, src);
        init(ts, i + (mid - tl) * 2, mid, tr, src);
        int value1 = ts->S[i + 1].value;
        int value2 = ts->S[i + (mid - tl) * 2].value;
        ts->S[i].value = magic(value1, value2);
    }
}

void tree_segment_init(TreeSegment* ts, int n, int def, int* src) {
    ts->n = n;
    ts->S = (Node*)malloc((2 * n + 1) * sizeof(Node));
    for (int i = 0; i < 2 * n + 1; i++) {
        ts->S[i].have_mod = 0;
    }
    init(ts, 0, 0, n, src);
}

int query(TreeSegment* ts, int i, int tl, int tr, int ql, int qr) {
    if (ql <= tl && tr <= qr) {
        return ts->S[i].value;
    } else {
        int mid = (tl + tr) / 2;
        if (tr <= ql || qr <= tl) {
            return 0;
        }
        int value1 = query(ts, i + 1, tl, mid, ql, qr);
        int value2 = query(ts, i + (mid - tl) * 2, mid, tr, ql, qr);
        return magic(value1, value2);
    }
}

void change(TreeSegment* ts, int i, int tl, int tr, int j, int _new) {
    if (tl + 1 == tr) {
        ts->S[i].value = _new;
    } else {
        int m = (tl + tr) / 2;
        if (j < m) {
            change(ts, ts->S[i].l, tl, m, j, _new);
        } else {
            change(ts, ts->S[i].r, m, tr, j, _new);
        }
        int value1 = ts->S[ts->S[i].l].value;
        int value2 = ts->S[ts->S[i].r].value;
        ts->S[i].value = magic(value1, value2);
    }
}

void update(TreeSegment* ts, int i, int tl, int tr, int ql, int qr, int mod) {
    if (ql <= tl && tr <= qr) {
        if (ts->S[i].have_mod) {
            ts->S[i].mod = merge_mods(ts->S[i].mod, mod, tl, tr);
        } else {
            ts->S[i].mod = mod;
            ts->S[i].have_mod = 1;
        }
    } else {
        int mid = (tl + tr) / 2;
        if (ql < mid) {
            update(ts, i + 1, tl, mid, ql, qr, mod);
        }
        if (mid < qr) {
            update(ts, i + (mid - tl) * 2, mid, tr, ql, qr, mod);
        }
        int value1 = ts->S[i + 1].value;
        int value2 = ts->S[i + (mid - tl) * 2].value;
        ts->S[i].value = magic(value1, value2);
    }
}

int find_right(TreeSegment* ts, int i, int tl, int tr, int L, int (*checker)(int), int pref) {
    if (tr <= L) {
        return -1;
    }
    if (tl + 1 == tr) {
        int _this = magic(pref, ts->S[i].value);
        if (checker(_this)) {
            return tl;
        } else {
            return -1;
        }
    } else {
        int mid = (tl + tr) / 2;
        if (L <= tl) {
            int _left = magic(pref, ts->S[i + 1].value);
            if (checker(_left)) {
                return find_right(ts, i + 1, tl, mid, L, checker, pref);
            } else {
                return find_right(ts, i + (mid - tl) * 2, mid, tr, L, checker, _left);
            }
        } else {
            int t = find_right(ts, i + 1, tl, mid, L, checker, pref);
            if (t != -1) {
                return t;
            }
            return find_right(ts, i + (mid - tl) * 2, mid, tr, L, checker, t);
        }
    }
}

int find_left(TreeSegment* ts, int i, int tl, int tr, int R, int (*checker)(int), int pref) {
    if (R < tl) {
        return -1;
    }
    if (tl + 1 == tr) {
        int _this = magic(pref, ts->S[i].value);
        if (checker(_this)) {
            return tl;
        } else {
            return -1;
        }
    } else {
        int mid = (tl + tr) / 2;
        if (tr <= R + 1) {
            int _right = magic(pref, ts->S[ts->S[i].r].value);
            if (checker(_right)) {
                return find_left(ts, ts->S[i].r, mid, tr, R, checker, pref);
            } else {
                return find_left(ts, ts->S[i].l, tl, mid, R, checker, _right);
            }
        } else {
            int t = find_left(ts, ts->S[i].r, mid, tr, R, checker, pref);
            if (t != -1) {
                return t;
            }
            return find_left(ts, ts->S[i].l, tl, mid, R, checker, t);
        }
    }
}

int main(int argc, char *argv[]) {
    int n, q;
    scanf("%d %d", &n, &q);
    int* pos = (int*)malloc(n * sizeof(int));
    int* w = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        scanf("%d", &pos[i]);
    }
    for (int i = 0; i < n; i++) {
        scanf("%d", &w[i]);
    }

    TreeSegment ts;
    tree_segment_init(&ts, n, 0, w);

    int* subw = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        subw[i] = (w[i] * (i - pos[i])) % MOD;
    }

    TreeSegment resp;
    tree_segment_init(&resp, n, 0, subw);

    for (int i = 0; i < q; i++) {
        int x, y;
        scanf("%d %d", &x, &y);
        if (x < 0) {
            int ind = -x;
            int _new = y;
            ind--;
            update(&ts, 0, 0, n, ind, ind + 1, _new);
            update(&resp, 0, 0, n, ind, ind + 1, (_new * (ind - pos[ind])) % MOD);
        } else {
            int l = x, r = y;
            l--;

            int sm = query(&ts, 0, 0, n, l, r);

            int bestL = find_right(&ts, 0, 0, n, l, [](int x) {
                return x * 2 >= sm;
            }, 0);

            int x = pos[bestL] - bestL + l;

            int left = 0;
            if (l < bestL) {
                left = query(&resp, 0, 0, n, l, bestL) + query(&ts, 0, 0, n, l, bestL) * (x - l);
            }
            int right = 0;
            if (bestL < r) {
                right = query(&resp, 0, 0, n, bestL, r) + query(&ts, 0, 0, n, bestL, r) * (x - l);
            }
            printf("%d\n", floor_mod(left - right, MOD));
        }
    }

    free(pos);
    free(w);
    free(subw);
    free(ts.S);
    free(resp.S);

    return 0;
}
